cluster/job_once_scheduler.go

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package cluster
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// syncedCallback uses the mutex to make things predictable for the client: the callback will be
+// called once at a time (the client does not need to worry about concurrency within the callback)
+type syncedCallback struct {
+	mu       sync.Mutex
+	callback func(string)
+}
+
+type syncedJobs struct {
+	mu   sync.RWMutex
+	jobs map[string]*JobOnce
+}
+
+type JobOnceScheduler struct {
+	pluginAPI JobPluginAPI
+
+	startedMu sync.RWMutex
+	started   bool
+
+	activeJobs     *syncedJobs
+	storedCallback *syncedCallback
+}
+
+var schedulerOnce sync.Once
+var s *JobOnceScheduler
+
+// GetJobOnceScheduler returns a scheduler which is ready to have its callback set. Repeated
+// calls will return the same scheduler.
+func GetJobOnceScheduler(pluginAPI JobPluginAPI) *JobOnceScheduler {
+	schedulerOnce.Do(func() {
+		s = &JobOnceScheduler{
+			pluginAPI: pluginAPI,
+			activeJobs: &syncedJobs{
+				jobs: make(map[string]*JobOnce),
+			},
+			storedCallback: &syncedCallback{},
+		}
+	})
+	return s
+}
+
+// Start starts the Scheduler. It finds all previous ScheduleOnce jobs and starts them running, and
+// fires any jobs that have reached or exceeded their runAt time. Thus, even if a cluster goes down
+// and is restarted, Start will restart previously scheduled jobs.
+func (s *JobOnceScheduler) Start() error {
+	s.startedMu.Lock()
+	defer s.startedMu.Unlock()
+	if s.started {
+		return errors.New("scheduler has already been started")
+	}
+
+	if err := s.verifyCallbackExists(); err != nil {
+		return errors.Wrap(err, "callback not found; cannot start scheduler")
+	}
+
+	if err := s.scheduleNewJobsFromDB(); err != nil {
+		return errors.Wrap(err, "could not start JobOnceScheduler due to error")
+	}
+
+	go s.pollForNewScheduledJobs()
+
+	s.started = true
+
+	return nil
+}
+
+// SetCallback sets the scheduler's callback. When a job fires, the callback will be called with
+// the job's id.
+func (s *JobOnceScheduler) SetCallback(callback func(string)) error {
+	if callback == nil {
+		return errors.New("callback cannot be nil")
+	}
+
+	s.storedCallback.mu.Lock()
+	defer s.storedCallback.mu.Unlock()
+
+	s.storedCallback.callback = callback
+	return nil
+}
+
+// ListScheduledJobs returns a list of the jobs in the db that have been scheduled. There is no
+// guarantee that list is accurate by the time the caller reads the list. E.g., the jobs in the list
+// may have been run, closed, or new jobs may have scheduled.
+func (s *JobOnceScheduler) ListScheduledJobs() ([]JobOnceMetadata, error) {
```

Comment:
Do we have measurements on how long this listing takes? I'm wondering if it would be better to have a dedicated SQL query that returned only the keys with the prefix (but maybe that's premature optimization).

Reply:
We used to do this all the time for the Jira plugin (with many many keys), and it's never been a problem. Not too worried about it because it's not a performance-critical piece of code. But sure we can always do it a bit faster with a SQL search... Maybe that would be a good plugin-api function!